src/hotspot/share/oops/instanceKlass.hpp

```diff
@@ -1430,4 +1430,58 @@ class InnerClassesIterator : public StackObj {
   }
 };
 
+// Iterator over class hierarchy under a particular class. Implements depth-first pre-order traversal.
+// Usage:
+//  for (ClassHierarchyIterator iter(root_klass); !iter.done(); iter.next()) {
+//    Klass* k = iter.klass();
+//    ...
+//  }
+class ClassHierarchyIterator : public StackObj {
+ private:
+  InstanceKlass* _root;
+  Klass*         _current;
+  bool           _visit_subclasses;
+
+ public:
+  ClassHierarchyIterator(InstanceKlass* root) : _root(root), _current(root), _visit_subclasses(true) {
+    assert(!root->is_interface(), "no subclasses");
+    assert(_root == _current, "required"); // initial state
+  }
+
+  bool done() {
+    return (_current == NULL);
+  }
+
+  // Make a step iterating over the class hierarchy under the root class.
+  // Skips subclasses if requested.
+  void next() {
+    assert(_current != NULL, "required");
+    if (_visit_subclasses && _current->subklass() != NULL) {
+      _current = _current->subklass();
+      return; // visit next subclass
+    }
+    _visit_subclasses = true; // reset
+    while (_current->next_sibling() == NULL && _current != _root) {
+      _current = _current->superklass(); // backtrack; no more sibling subclasses left
+    }
+    if (_current == _root) {
+      // Iteration is over (back at root after backtracking). Invalidate the iterator.
+      _current = NULL;
+      return;
+    }
+    _current = _current->next_sibling();
+    return; // visit next sibling subclass
+  }
+
+  Klass* klass() {
+    assert(!done(), "sanity");
+    return _current;
+  }
+
+  // Skip subclasses of the current class.
+  void skip_subclasses() {
+    _visit_subclasses = false;
+  }
+};
+
```

Comment:
The _subklass and _next_sibling fields and implementation are in Klass (klass.hpp/cpp) but I always wonder why they are not in InstanceKlass (instanceKlass.hpp/cpp).  If this new class is in instanceKlass.hpp, these fields should be in the same.  If you agree, we should file an RFE and I will move them.  If the fields truly belong in klass.hpp, then this should be there too. ie. they should match.

Reply:
I think it's `java.lang.Object` which complicates things. All array classes are rooted at Object, so neither `_subklass` nor  `_next_sibling` can be changed to `InstanceKlass`. I put `ClassHierarchyIterator` in `instanceKlass.hpp` primarily because it accepts only `InstanceKlass` root class. But I'm fine with putting it into `klass.hpp`.

Reply:
Does the compiler dependencies use the _subklass lists for array classes?  The only other use for this is vtable reinitialization.  It is fine to leave ClassHIerarchyIterator in instanceKlass.hpp since that's how it's used in this change. I might investigate this more some later time.

Reply:
> Does the compiler dependencies use the _subklass lists for array classes?

Arrays aren't interesting in the context of CHA (which Dependencies implements), so they are handled only because they are encountered during hierarchy traversal under `Object`:
```
  bool is_witness(Klass* k) {
    if (doing_subtype_search()) {
      return Dependencies::is_concrete_klass(k);
    } else if (!k->is_instance_klass()) {
      return false; // no methods to find in an array type
    } else {
      ...
```